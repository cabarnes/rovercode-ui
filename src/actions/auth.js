// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const UPDATE_VALID_AUTH = 'UPDATE_VALID_AUTH';
export const USER_LOGOUT = 'USER_LOGOUT';

// action creators
export const updateValidAuth = (isValidAuth) => ({
  type: UPDATE_VALID_AUTH,
  payload: isValidAuth,
});

export const logout = () => ({
  type: USER_LOGOUT,
});

// helper functions
export const checkAuthError = (dispatch) => (error) => {
  if (error.response && error.response.status === 401) {
    // Authentication is no longer valid
    dispatch(updateValidAuth(false));
  } else {
    throw error;
  }
};

export const authHeader = (cookies) => ({
  headers: {
    Authorization: `JWT ${cookies.get('auth_jwt')}`,
  },
});
