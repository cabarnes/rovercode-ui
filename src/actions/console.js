// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const APPEND = 'APPEND';
export const CLEAR = 'CLEAR';

// action creators
export const append = (message) => ({
  type: APPEND,
  payload: message,
});

export const clear = () => ({
  type: CLEAR,
});
