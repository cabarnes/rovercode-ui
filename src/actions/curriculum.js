// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_COURSES = 'FETCH_COURSES';
export const FETCH_COURSES_PENDING = `${FETCH_COURSES}_PENDING`;
export const FETCH_COURSES_FULFILLED = `${FETCH_COURSES}_FULFILLED`;
export const FETCH_COURSES_REJECTED = `${FETCH_COURSES}_REJECTED`;

// action creators
export const fetchCourses = (xhrOptions) => ({
  type: FETCH_COURSES,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/courses/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});
