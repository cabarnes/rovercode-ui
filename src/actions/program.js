// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const CLEAR_PROGRAMS = 'CLEAR_PROGRAMS';
export const FETCH_PROGRAMS = 'FETCH_PROGRAMS';
export const FETCH_PROGRAMS_PENDING = `${FETCH_PROGRAMS}_PENDING`;
export const FETCH_PROGRAMS_FULFILLED = `${FETCH_PROGRAMS}_FULFILLED`;
export const FETCH_PROGRAMS_REJECTED = `${FETCH_PROGRAMS}_REJECTED`;
export const REMOVE_PROGRAM = 'REMOVE_PROGRAM';
export const REMOVE_PROGRAM_PENDING = `${REMOVE_PROGRAM}_PENDING`;
export const REMOVE_PROGRAM_FULFILLED = `${REMOVE_PROGRAM}_FULFILLED`;
export const REMOVE_PROGRAM_REJECTED = `${REMOVE_PROGRAM}_REJECTED`;

// action creators
export const clearPrograms = () => ({
  type: CLEAR_PROGRAMS,
});

export const fetchPrograms = (xhrOptions) => ({
  type: FETCH_PROGRAMS,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/block-diagrams/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const removeProgram = (id, xhrOptions) => ({
  type: REMOVE_PROGRAM,
  payload: axios.delete(`/api/v1/block-diagrams/${id}/`, xhrOptions)
    .then(({ data }) => (
      data
    )),
});
