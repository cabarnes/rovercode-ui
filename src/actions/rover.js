// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const SCAN = 'SCAN';
export const SCAN_FULFILLED = `${SCAN}_FULFILLED`;
export const SCAN_REJECTED = `${SCAN}_REJECTED`;
export const CONNECT_ROVER = 'CONNECT_ROVER';
export const CONNECT_ROVER_FULFILLED = `${CONNECT_ROVER}_FULFILLED`;
export const CONNECT_ROVER_REJECTED = `${CONNECT_ROVER}_REJECTED`;
export const SEND_ROVER = 'SEND_ROVER';
export const SEND_ROVER_FULFILLED = `${SEND_ROVER}_FULFILLED`;
export const SEND_ROVER_REJECTED = `${SEND_ROVER}_REJECTED`;
export const DISCONNECT_ROVER = 'DISCONNECT_ROVER';

const UART_SERVICE_UUID = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
const UART_TX_CHARACTERISTIC_UUID = '6e400002-b5a3-f393-e0a9-e50e24dcca9e';
const UART_RX_CHARACTERISTIC_UUID = '6e400003-b5a3-f393-e0a9-e50e24dcca9e';

// action creators
export const scan = () => ({
  type: SCAN,
  payload: navigator.bluetooth.requestDevice({
    filters: [{ namePrefix: 'BBC micro:bit' }],
    optionalServices: [UART_SERVICE_UUID],
  }),
});

export const connect = (device, onMessage) => ({
  type: CONNECT_ROVER,
  payload: device.gatt.connect()
    .then(server => server.getPrimaryService(UART_SERVICE_UUID))
    .then(service => Promise.all([
      service.getCharacteristic(UART_RX_CHARACTERISTIC_UUID),
      service.getCharacteristic(UART_TX_CHARACTERISTIC_UUID)
        .then((characteristic) => {
          characteristic.startNotifications();
          characteristic.addEventListener(
            'characteristicvaluechanged',
            onMessage,
          );
          return characteristic;
        }),
    ])),
});

export const send = (channel, message) => ({
  type: SEND_ROVER,
  payload: channel.writeValue(message),
});

export const disconnect = device => ({
  type: DISCONNECT_ROVER,
  payload: device.gatt.disconnect(),
});
