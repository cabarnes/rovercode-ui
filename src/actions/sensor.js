// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const CHANGE_LEFT_SENSOR_STATE = 'CHANGE_LEFT_SENSOR_STATE';
export const CHANGE_RIGHT_SENSOR_STATE = 'CHANGE_RIGHT_SENSOR_STATE';
export const CHANGE_LIGHT_SENSOR_READINGS = 'CHANGE_LIGHT_SENSOR_READINGS';

// Sensor States
export const COVERED = 1;
export const NOT_COVERED = 2;

// action creators
export const changeLeftSensorState = (state) => ({
  type: CHANGE_LEFT_SENSOR_STATE,
  payload: state,
});

export const changeRightSensorState = (state) => ({
  type: CHANGE_RIGHT_SENSOR_STATE,
  payload: state,
});

export const changeLightSensorReadings = (leftReading, rightReading) => ({
  type: CHANGE_LIGHT_SENSOR_READINGS,
  payload: { leftReading, rightReading },
});
