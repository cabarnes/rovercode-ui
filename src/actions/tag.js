// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_TAGS = 'FETCH_TAGS';
export const FETCH_TAGS_PENDING = `${FETCH_TAGS}_PENDING`;
export const FETCH_TAGS_FULFILLED = `${FETCH_TAGS}_FULFILLED`;
export const FETCH_TAGS_REJECTED = `${FETCH_TAGS}_REJECTED`;

// action creators
export const fetchTags = (xhrOptions) => ({
  type: FETCH_TAGS,
  payload: axios.get('/api/v1/tags/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});
