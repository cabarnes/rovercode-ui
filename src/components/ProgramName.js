import React, { Component } from 'react';
import { Save } from '@material-ui/icons';
import {
  IconButton,
  Input,
  InputAdornment,
  InputLabel,
} from '@material-ui/core';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies } from 'react-cookie';
import { injectIntl } from 'react-intl';
import PropTypes from 'prop-types';

import { checkAuthError, authHeader } from '@/actions/auth';
import { changeName as actionChangeName } from '@/actions/code';

const mapStateToProps = ({ code }) => ({ code });
const mapDispatchToProps = (dispatch, { cookies }) => ({
  changeName: (id, name) => dispatch(actionChangeName(id, name, authHeader(cookies)))
    .catch(checkAuthError(dispatch)),
});

class ProgramName extends Component {
  constructor(props) {
    super(props);

    this.state = {
      editingName: '',
      previousPropName: null, // eslint-disable-line react/no-unused-state
    };
  }

  static getDerivedStateFromProps(props, state) {
    const { code } = props;
    const { previousPropName } = state;

    if (code.name !== previousPropName) {
      return {
        editingName: code.name,
        previousPropName: code.name,
      };
    }

    return null;
  }

  handleChange = (e) => {
    this.setState({
      editingName: e.target.value,
    });
  }

  handleSave = () => {
    const { changeName, code } = this.props;
    const { editingName } = this.state;

    changeName(code.id, editingName);
  }

  saveAdornment = () => {
    const { editingName, previousPropName } = this.state;

    return editingName !== previousPropName ? (
      <InputAdornment position="end">
        <IconButton color="secondary" onClick={this.handleSave}>
          <Save />
        </IconButton>
      </InputAdornment>
    ) : (null);
  }

  render() {
    const { code, intl } = this.props;
    const { editingName } = this.state;

    const nameLabel = intl.formatMessage({
      id: 'app.program_name.name',
      description: 'Label for new program name entry',
      defaultMessage: 'Name:',
    });

    return (
      <>
        <InputLabel htmlFor="program-name">{nameLabel}</InputLabel>
        <Input
          id="program-name"
          value={editingName}
          disabled={code.isReadOnly}
          onChange={this.handleChange}
          endAdornment={this.saveAdornment()}
        />
      </>
    );
  }
}

ProgramName.propTypes = {
  code: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    isReadOnly: PropTypes.bool,
  }).isRequired,
  changeName: PropTypes.func.isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default hot(module)(
  withCookies(
    injectIntl(
      connect(mapStateToProps, mapDispatchToProps)(ProgramName),
    ),
  ),
);
