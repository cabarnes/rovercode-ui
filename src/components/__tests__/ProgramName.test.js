import React from 'react';
import { Input } from '@material-ui/core';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Cookies } from 'react-cookie';

import { updateValidAuth } from '@/actions/auth';
import { changeName } from '@/actions/code';
import ProgramName from '../ProgramName';

const cookiesValues = { auth_jwt: '1234' };
const cookies = new Cookies(cookiesValues);

describe('The ProgramName component', () => {
  const mockStore = configureStore();
  const context = { cookies };
  let store;

  beforeEach(() => {
    store = mockStore({
      code: {
        name: 'test name',
        isReadOnly: false,
      },
    });
    store.dispatch = jest.fn(() => Promise.resolve());
  });

  test('renders on the page with no errors', () => {
    const wrapper = mountWithIntl(<ProgramName store={store} />, { context });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('displays name', () => {
    const wrapper = shallowWithIntl(
      <ProgramName store={store} />, { context },
    ).dive().dive().dive()
      .dive()
      .dive()
      .dive()
      .dive();

    expect(wrapper.find(Input).length).toBe(1);
    expect(wrapper.find(Input).props().value).toBe('test name');
    expect(wrapper.find(Input).props().disabled).toBe(false);
  });

  test('disabled when read only', () => {
    const localStore = mockStore({
      code: {
        name: 'test name',
        isReadOnly: true,
      },
    });
    const wrapper = shallowWithIntl(
      <ProgramName store={localStore} />,
      { context },
    ).dive().dive().dive()
      .dive()
      .dive()
      .dive()
      .dive();

    expect(wrapper.find(Input).props().disabled).toBe(true);
    expect(wrapper.find(Input).props().endAdornment).toBeNull();
  });

  test('handles change', () => {
    const wrapper = shallowWithIntl(
      <ProgramName store={store} />, { context },
    ).dive().dive().dive()
      .dive()
      .dive()
      .dive()
      .dive();

    wrapper.find(Input).simulate('change', { target: { value: 'new name' } });
    wrapper.update();

    expect(wrapper.find(Input).props().value).toBe('new name');
    expect(wrapper.find(Input).props().endAdornment).toBeDefined();
  });

  test('handles save', () => {
    const wrapper = shallowWithIntl(
      <ProgramName store={store} />, { context },
    ).dive().dive().dive()
      .dive()
      .dive()
      .dive()
      .dive();

    wrapper.find(Input).simulate('change', { target: { value: 'new name' } });
    wrapper.update();

    wrapper.instance().handleSave();

    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(changeName('test name'));
  });

  test('handles authentication error', (done) => {
    const error = new Error();
    error.response = {
      status: 401,
    };
    store.dispatch = jest.fn();
    store.dispatch.mockReturnValueOnce(Promise.reject(error));
    store.dispatch.mockReturnValue(Promise.resolve());

    const wrapper = shallowWithIntl(<ProgramName store={store} />, { context }).dive();
    wrapper.dive().dive().dive().dive()
      .dive()
      .props()
      .changeName(1, 'testname')
      .then(() => {
        expect(store.dispatch.mock.calls.length).toBe(2);
        expect(store.dispatch).toHaveBeenCalledWith(
          changeName(1, 'testname', {
            headers: {
              Authorization: `JWT ${cookiesValues.auth_jwt}`,
            },
          }),
        );
        expect(store.dispatch).toHaveBeenCalledWith(updateValidAuth(false));
        done();
      });
  });

  test('handles other error', (done) => {
    const error = new Error();
    error.response = {
      status: 500,
    };
    store.dispatch = jest.fn(() => Promise.reject(error));

    const wrapper = shallowWithIntl(<ProgramName store={store} />, { context }).dive();
    wrapper.dive().dive().dive().dive()
      .dive()
      .props()
      .changeName(1, 'testname')
      .catch(() => {
        expect(store.dispatch.mock.calls.length).toBe(1);
        expect(store.dispatch).toHaveBeenCalledWith(
          changeName(1, 'testname', {
            headers: {
              Authorization: `JWT ${cookiesValues.auth_jwt}`,
            },
          }),
        );
        done();
      });
  });
});
