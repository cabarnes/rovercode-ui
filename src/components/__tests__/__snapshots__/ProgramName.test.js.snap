// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ProgramName component renders on the page with no errors 1`] = `
<withCookies(injectIntl(Connect(ProgramName)))
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <withCookies(injectIntl(Connect(ProgramName)))
    cookies={
      Cookies {
        "HAS_DOCUMENT_COOKIE": true,
        "changeListeners": Array [
          [Function],
        ],
        "cookies": Object {},
      }
    }
    forwardedRef={null}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <injectIntl(Connect(ProgramName))
      allCookies={Object {}}
      cookies={
        Cookies {
          "HAS_DOCUMENT_COOKIE": true,
          "changeListeners": Array [
            [Function],
          ],
          "cookies": Object {},
        }
      }
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(ProgramName)
        allCookies={Object {}}
        cookies={
          Cookies {
            "HAS_DOCUMENT_COOKIE": true,
            "changeListeners": Array [
              [Function],
            ],
            "cookies": Object {},
          }
        }
        intl={
          Object {
            "defaultFormats": Object {},
            "defaultLocale": "en",
            "formatDate": [Function],
            "formatDateToParts": [Function],
            "formatDisplayName": [Function],
            "formatHTMLMessage": [Function],
            "formatList": [Function],
            "formatMessage": [Function],
            "formatNumber": [Function],
            "formatNumberToParts": [Function],
            "formatPlural": [Function],
            "formatRelativeTime": [Function],
            "formatTime": [Function],
            "formatTimeToParts": [Function],
            "formats": Object {},
            "formatters": Object {
              "getDateTimeFormat": [Function],
              "getDisplayNames": [Function],
              "getListFormat": [Function],
              "getMessageFormat": [Function],
              "getNumberFormat": [Function],
              "getPluralRules": [Function],
              "getRelativeTimeFormat": [Function],
            },
            "locale": "en",
            "messages": Object {},
            "onError": [Function],
            "textComponent": Symbol(react.fragment),
            "timeZone": undefined,
          }
        }
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [MockFunction],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <ProgramName
          allCookies={Object {}}
          changeName={[Function]}
          code={
            Object {
              "isReadOnly": false,
              "name": "test name",
            }
          }
          cookies={
            Cookies {
              "HAS_DOCUMENT_COOKIE": true,
              "changeListeners": Array [
                [Function],
              ],
              "cookies": Object {},
            }
          }
          intl={
            Object {
              "defaultFormats": Object {},
              "defaultLocale": "en",
              "formatDate": [Function],
              "formatDateToParts": [Function],
              "formatDisplayName": [Function],
              "formatHTMLMessage": [Function],
              "formatList": [Function],
              "formatMessage": [Function],
              "formatNumber": [Function],
              "formatNumberToParts": [Function],
              "formatPlural": [Function],
              "formatRelativeTime": [Function],
              "formatTime": [Function],
              "formatTimeToParts": [Function],
              "formats": Object {},
              "formatters": Object {
                "getDateTimeFormat": [Function],
                "getDisplayNames": [Function],
                "getListFormat": [Function],
                "getMessageFormat": [Function],
                "getNumberFormat": [Function],
                "getPluralRules": [Function],
                "getRelativeTimeFormat": [Function],
              },
              "locale": "en",
              "messages": Object {},
              "onError": [Function],
              "textComponent": Symbol(react.fragment),
              "timeZone": undefined,
            }
          }
          store={
            Object {
              "clearActions": [Function],
              "dispatch": [MockFunction],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            }
          }
        >
          <Input
            disabled={false}
            label="Name:"
            onChange={[Function]}
            type="text"
            value="test name"
          >
            <div
              className="ui labeled input"
            >
              <Label
                className="label"
                content="Name:"
              >
                <div
                  className="ui label label"
                  onClick={[Function]}
                >
                  Name:
                </div>
              </Label>
              <input
                disabled={false}
                onChange={[Function]}
                type="text"
                value="test name"
              />
            </div>
          </Input>
          <Confirm
            cancelButton="Cancel"
            confirmButton="OK"
            content="Are you sure that you want to change the name of this program?"
            onCancel={[Function]}
            onConfirm={[Function]}
            open={false}
            size="small"
          >
            <Modal
              centered={true}
              closeOnDimmerClick={true}
              closeOnDocumentClick={false}
              dimmer={true}
              eventPool="Modal"
              onClose={[Function]}
              open={false}
              size="small"
            >
              <Portal
                closeOnDocumentClick={false}
                closeOnEscape={true}
                eventPool="Modal"
                mountNode={<body />}
                onClose={[Function]}
                onMount={[Function]}
                onOpen={[Function]}
                onUnmount={[Function]}
                open={false}
                openOnTriggerClick={true}
              />
            </Modal>
          </Confirm>
        </ProgramName>
      </Connect(ProgramName)>
    </injectIntl(Connect(ProgramName))>
  </withCookies(injectIntl(Connect(ProgramName)))>
</withCookies(injectIntl(Connect(ProgramName)))>
`;
